ambari-server/src/main/java/org/apache/ambari/server/events/MessageEmitter.java

```diff
@@ -17,70 +17,248 @@
+      new ThreadFactoryBuilder().setNameFormat("agent-message-emitter-%d").build());
+  protected final ExecutorService monitorsExecutor = Executors.newFixedThreadPool(10,
+      new ThreadFactoryBuilder().setNameFormat("ambari-message-monitor-%d").build());
+
+  protected static final AtomicLong MESSAGE_ID = new AtomicLong(0);
+  protected ConcurrentHashMap<Long, ScheduledFuture> unconfirmedMessages = new ConcurrentHashMap<>();
+  protected ConcurrentHashMap<Long, BlockingQueue<ExecutionCommandEvent>> messagesToEmit = new ConcurrentHashMap<>();
+
+  // is used to cancel agent queue check on unregistering
+  protected ConcurrentHashMap<Long, Future> monitors = new ConcurrentHashMap<>();
 
+  public final int retryCount;
+  public final int retryInterval;
 
-  public MessageEmitter(AgentSessionManager agentSessionManager, SimpMessagingTemplate simpMessagingTemplate) {
+  public MessageEmitter(AgentSessionManager agentSessionManager, SimpMessagingTemplate simpMessagingTemplate,
+                        AmbariEventPublisher ambariEventPublisher, int retryCount, int retryInterval) {
     this.agentSessionManager = agentSessionManager;
     this.simpMessagingTemplate = simpMessagingTemplate;
+    this.ambariEventPublisher = ambariEventPublisher;
+    this.retryCount = retryCount;
+    this.retryInterval = retryInterval;
+    ambariEventPublisher.register(this);
   }
 
   /**
    * Determines destinations and emits message.
    * @param event message should to be emitted.
    * @throws AmbariException
    */
-  abstract void emitMessage(STOMPEvent event) throws AmbariException;
+  abstract void emitMessage(STOMPEvent event) throws AmbariException, InterruptedException;
+
+  public void emitMessageRetriable(ExecutionCommandEvent event) throws AmbariException, InterruptedException {
+    // set message identifier used to recognize NACK/ACK agent response
+    event.setMessageId(MESSAGE_ID.getAndIncrement());
+
+    Long hostId = event.getHostId();
+    if (!messagesToEmit.containsKey(hostId)) {
+      LOG.error("Trying to emit message to unregistered host with id {}", hostId);
+      return;
+    }
+    messagesToEmit.get(hostId).add(event);
+  }
+
+  private class MessagesToEmitMonitor implements Runnable {
+
+    private final Long hostId;
+
+    public MessagesToEmitMonitor(Long hostId) {
+      this.hostId = hostId;
+    }
+
+    @Override
+    public void run() {
+      while (true) {
+        try {
+          ExecutionCommandEvent event = messagesToEmit.get(hostId).take();
+          EmitMessageTask emitMessageTask = new EmitMessageTask(event);
+          ScheduledFuture scheduledFuture =
+              emittersExecutor.scheduleAtFixedRate(emitMessageTask,
+                  0, retryInterval, TimeUnit.SECONDS);
+          emitMessageTask.setScheduledFuture(scheduledFuture);
+          unconfirmedMessages.put(event.getMessageId(), scheduledFuture);
+
+          scheduledFuture.get();
```

Comment:
Can we afford to wait without timeout here? The monitors could get exhausted, right?